File: skript-parser-java/src/main/java/io/github/syst3ms/skriptparser/classes/Literal.java
package io.github.syst3ms.skriptparser.classes;

import java.lang.reflect.Array;
import java.util.Arrays;

public class Literal<T> implements Expression<T> {
    private T[] values;

    @SuppressWarnings("unchecked")
    @SafeVarargs
    public Literal(Class<T> c, T... values) {
        this.values = (T[]) Array.newInstance(c, values.length);
        System.arraycopy(values, 0, this.values, 0, values.length);
    }

    @Override
    public boolean init(Expression<?>[] expressions, int matchedPattern, ParseResult parseResult) {
        return true;
    }

    @Override
    public T[] getValues() {
        return values;
    }

    @Override
    public String toString(boolean debug) {
        if (isSingle()) {
            return values[0].toString();
        } else {
            return Arrays.toString(values); // For now TODO make a proper way of doing this
        }
    }

    // This is only necessary here, as literals aren't registered per-say
    public boolean isSingle() {
        return values.length == 1;
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/classes/Literal.kt
package io.github.syst3ms.skriptparser.classes

import java.util.Arrays

class Literal<T>(c: Class<T>, vararg values: T) : Expression<T> {
    private var values: Array<T>

    init {
        @Suppress("UNCHECKED_CAST")
        val array : Array<T> = java.lang.reflect.Array.newInstance(c, values.size) as Array<T>
        this.values = array
    }

    // This is only necessary here, as literals aren't registered per-say
    val isSingle: Boolean
        get() = values.size == 1

    override fun init(expressions: Array<Expression<out Any>>, matchedPattern: Int, parseResult: ParseResult) = true

    override fun getValues() = values

    override fun toString(debug: Boolean) = if (isSingle) {
        values[0].toString()
    } else {
        Arrays.toString(values) // For now TODO make a proper way of doing this
    }
}
